domain/storage/service/storage_test.go

```diff
@@ -0,0 +1,281 @@
+	provider           *MockProvider
+	volumeSource       *MockVolumeSource
+	volumeImporter     *MockVolumeImporter
+	filesystemSource   *MockFilesystemSource
+	filesystemImporter *MockFilesystemImporter
+}
+
+var _ = gc.Suite(&storageSuite{})
+
+type volumeImporter struct {
+	*MockVolumeSource
+	*MockVolumeImporter
+}
+
+type filesystemImporter struct {
+	*MockFilesystemSource
+	*MockFilesystemImporter
+}
+
+func (s *storageSuite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+
+	s.state = NewMockState(ctrl)
+	s.volumeSource = NewMockVolumeSource(ctrl)
+	s.volumeImporter = NewMockVolumeImporter(ctrl)
+	s.filesystemSource = NewMockFilesystemSource(ctrl)
+	s.filesystemImporter = NewMockFilesystemImporter(ctrl)
+	s.provider = NewMockProvider(ctrl)
+
+	registry := NewMockProviderRegistry(ctrl)
+	registry.EXPECT().StorageProvider(storage.ProviderType("ebs")).Return(s.provider, nil).AnyTimes()
+	registry.EXPECT().StorageProvider(storage.ProviderType("elastic")).Return(s.provider, nil).AnyTimes()
+
+	s.registry = storage.ChainedProviderRegistry{registry, provider.CommonStorageProviders()}
+
+	return ctrl
+}
+
+func (s *storageSuite) service(c *gc.C) *Service {
+	return NewService(s.state, loggertesting.WrapCheckLog(c), modelStorageRegistryGetter(func() storage.ProviderRegistry {
+		return s.registry
+	}))
+}
+
+func (s *storageSuite) TestImportFilesystemValidate(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	_, err := s.service(c).ImportFilesystem(context.Background(), ImportStorageParams{
+		Kind:        storage.StorageKindFilesystem,
+		Pool:        "elastic",
+		ProviderId:  "provider-id",
+		StorageName: "0",
+	})
+	c.Assert(err, jc.ErrorIs, corestorage.InvalidStorageName)
+
+	_, err = s.service(c).ImportFilesystem(context.Background(), ImportStorageParams{
+		Kind:        storage.StorageKindFilesystem,
+		Pool:        "0",
+		ProviderId:  "provider-id",
+		StorageName: "pgdata",
+	})
+	c.Assert(err, jc.ErrorIs, storageerrors.InvalidPoolNameError)
+
+	_, err = s.service(c).ImportFilesystem(context.Background(), ImportStorageParams{
+		Kind:        storage.StorageKindBlock,
+		Pool:        "elastic",
+		ProviderId:  "provider-id",
+		StorageName: "pgdata",
+	})
+	c.Assert(err, jc.ErrorIs, errors.NotSupported)
+}
+
+func (s *storageSuite) TestImportFilesystem(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.provider.EXPECT().Supports(storage.StorageKindFilesystem).Return(true)
+	cfg, err := storage.NewConfig("elastic", "elastic", nil)
+	c.Assert(err, jc.ErrorIsNil)
+	s.provider.EXPECT().FilesystemSource(cfg).Return(filesystemImporter{
+		MockFilesystemSource:   s.filesystemSource,
+		MockFilesystemImporter: s.filesystemImporter,
+	}, nil)
+
+	s.state.EXPECT().GetStoragePoolByName(gomock.Any(), "elastic").Return(domainstorage.StoragePoolDetails{}, storageerrors.PoolNotFoundError)
+	s.state.EXPECT().GetModelDetails().Return(domainstorage.ModelDetails{
+		ModelUUID:      coretesting.ModelTag.Id(),
```

Comment:
I don't think we should be bringing tag's down here. Can we just gen these once in setup and save them on the suite ?

Reply:
This is the universal pattern we've been using in every test package to date. Same with controller UUID. The pattern goes back years. If we want to change it, we could add something to the core testing package alongside the model and controller tags. Generating and saving attrs on the test suite works but is unnecessary overhead.

Reply:
For these tests, I've used generated uuids instead of the testing ones. We can look to holistically adopt a consistent approach over time.